File: mobile/shared/src/commonTest/kotlin/uk/foodclassifier/IngredientTextTest.kt
package uk.foodclassifier

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue
import uk.foodclassifier.text.IngredientLabelParser
import uk.foodclassifier.text.OcrNormalizer

class IngredientTextTest {
    @Test
    fun splitRespectingBracketsKeepsNestedStructure() {
        val parts = splitRespectingBrackets(
            "Protein Blend [Calcium Caseinate (Milk), Whey Protein Isolate (Milk)]",
        )
        assertEquals(1, parts.size)
        assertTrue(parts[0].contains("["))
        assertTrue(parts[0].contains("(Milk)"))
    }

    @Test
    fun extractsMilkFromSquareAndRoundBrackets() {
        val raw =
            "Protein Blend [Calcium Caseinate (Milk), Whey Protein Isolate (Milk)]"
        val segments = ingredientSegmentsForMatching(raw)
        assertTrue(segments.any { it.contains("milk") })
    }

    @Test
    fun repairsOcrMilkArtifacts() {
        val segments = ingredientSegmentsForMatching(
            "Whey Protein Isolate lMilk, Calcium Caseinate Mi ME-NGR",
        )
        assertTrue(segments.any { it.contains("milk") })
    }

    @Test
    fun structuralRepairKeepsWordSpacing() {
        val input = "Reduced Fat Soft C(hese (Milk), Salt Stablisers: Guar Gum"
        val structural = OcrNormalizer.normalize(input)
        assertEquals(
            "Reduced Fat Soft Cheese (Milk), Salt, Stablisers: Guar Gum",
            structural,
        )
    }

    @Test
    fun parserFormatsIngredientsOnePerLine() {
        assertEquals(
            "Reduced Fat Soft Cheese (Milk)\nSalt\nGuar Gum",
            IngredientLabelParser.parseBlock(
                "Reduced Fat Soft C(hese (Milk), Salt Stablisers: Guar Gum",
            ).displayText,
        )
    }
}

File: flutter/test/ingredient_text_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:food_classifier/food_classifier.dart';

void main() {
  test('splitRespectingBracketsKeepsNestedStructure', () {
    final parts = splitRespectingBrackets(
      'Protein Blend [Calcium Caseinate (Milk), Whey Protein Isolate (Milk)]',
    );
    expect(parts, hasLength(1));
    expect(parts.first.contains('['), isTrue);
    expect(parts.first.contains('(Milk)'), isTrue);
  });

  test('extractsMilkFromSquareAndRoundBrackets', () {
    const raw =
        'Protein Blend [Calcium Caseinate (Milk), Whey Protein Isolate (Milk)]';
    final segments = ingredientSegmentsForMatching(raw);
    expect(segments.any((segment) => segment.contains('milk')), isTrue);
  });

  test('repairsOcrMilkArtifacts', () {
    final segments = ingredientSegmentsForMatching(
      'Whey Protein Isolate lMilk, Calcium Caseinate Mi ME-NGR',
    );
    expect(segments.any((segment) => segment.contains('milk')), isTrue);
  });

  test('structuralRepairKeepsWordSpacing', () {
    const input = 'Reduced Fat Soft C(hese (Milk), Salt Stablisers: Guar Gum';
    final structural = OcrNormalizer.normalize(input);
    expect(
      structural,
      'Reduced Fat Soft Cheese (Milk), Salt, Stablisers: Guar Gum',
    );
  });

  test('detectsGarbledStabilisersAsFunctionalLabel', () {
    expect(isFunctionalLabel('Sabilisers'), isTrue);
    expect(
      expandFunctionalLabelToIngredients('Sabilisers: Guar Gum'),
      ['Guar Gum'],
    );
  });

  test('detectsGarbledAcidityRegulatorAsFunctionalLabel', () {
    expect(isFunctionalLabel('Acidity Requlator'), isTrue);
    expect(
      expandFunctionalLabelToIngredients('Acidity Requlator: Sodium Hydroxide'),
      ['Sodium Hydroxide'],
    );
  });

  test('expandsFunctionalLabelWithParentheses', () {
    expect(
      expandFunctionalLabelToIngredients(
        'sweetener (steviol glycosides from stevia)',
      ),
      ['steviol glycosides from stevia'],
    );
  });

  test('parserFormatsIngredientsOnePerLine', () {
    expect(
      IngredientLabelParser.parseBlock(
        'Reduced Fat Soft C(hese (Milk), Salt Stablisers: Guar Gum',
      ).displayText,
      'Reduced Fat Soft Cheese (Milk)\nSalt\nGuar Gum',
    );
  });
}
